### PR TITLE
API: use equals_identical for __eq__

### LINKED ISSUE #1732: REGR: equality (__eq__) now ignores the z dimension
With shapely 2.0, we switched the implementation of `Geometry.__eq__` to use the "equals_exact" predicate from GEOS, instead of manually comparing the coordinates as we did before (`geom1.coords == geom2.coords`), in the assumption this is equivalent. But it seems that `GEOSEqualsExact` only checks the X and Y coordinates (https://github.com/libgeos/geos/issues/809), thus introducing a regression in comparing 3D geometries:

With shapely 1.8
```python
>>> Point(1, 2, 3) == Point(1, 2, 4)
False
```

With shapely 2.0:

```python
>>> Point(1, 2, 3) == Point(1, 2, 4)
True
```

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #1732, closes https://github.com/shapely/shapely/issues/1775

This might be a bit overkill for _just_ implementing `__eq__`, but this should also allow to actually expose "equals_identical" as a ufunc for all GEOS versions (and not only for GEOS>=3.12). And it is also faster than the easier python alternative (but not sure how significant this will be in real-world cases):

```
geom = shapely.LineString(np.random.randn(1000, 2))

In [19]: %timeit geom == geom
10.8 µs ± 198 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [20]: %timeit np.allclose(shapely.get_coordinates(geom), shapely.get_coordinates(geom), rtol=0, equal_nan=True)
76.5 µs ± 599 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Additional complication: I forgot that `GEOSCoordSeq_copyToBuffer_r` is only available for more recent GEOS versions (>= 3.10), so if we want to go this way, we need a fallback for older versions (we actually do already have a similar fallback implemented for the equivalent `GEOSCoordSeq_copyFromBuffer_r`). 

The other reason that some tests are failing is because this PR actually changes behaviour regarding NaNs. On main, we don't regard such geometries as equal:

```
In [2]: import shapely

In [3]: line_string_nan = shapely.LineString([(np.nan, np.nan), (np.nan, np.nan)])

In [4]: line_string_nan
Out[4]: <LINESTRING (NaN NaN, NaN NaN)>

In [5]: line_string_nan == line_string_nan
Out[5]: False
```

while with this PR, I made NaNs in the same location as equal (following what was done for GEOSEqualsIdentical upstream in  https://github.com/libgeos/geos/pull/810). 

With shapely 1.x, this also resulted in False, because we implemented `__eq__` as `tuple(self.coords) == tuple(other.coords)`, which ends up comparing the scalar coordinate values, and then `nan != nan`.

**Comment by idanmiara** [comment]
@jorisvandenbossche Thanks for fixing this issue!

**Review by brendan-ward — COMMENTED** [review]
Thanks @jorisvandenbossche 

A few relatively minor comments on this so far; assume that this is still WIP waiting on `GEOSCoordSeq_copyToBuffer_r`.

Should this use `GEOSEqualsIdentical` if available from GEOS and skip the custom implementation here?

**Comment by jorisvandenbossche** [comment]
@brendan-ward thanks a lot for the detailed review. 
Higher level question: does that mean you would be OK with the behaviour change to consider NaNs in the same location in the coordinate sequence as equal? Because that's a deviation from what `__eq__` does now, so something we should decide consciously.

**Comment by idanmiara** [comment]
> @brendan-ward thanks a lot for the detailed review. Higher level question: does that mean you would be OK with the behaviour change to consider NaNs in the same location in the coordinate sequence as equal? Because that's a deviation from what `__eq__` does now, so something we should decide consciously.

For me the proposed behavior makes much more sense, that NaNs would be considered equal.
Say for instance we can compare safely 3D points that happen to have `Z=NaN` `assert Point(1,2,NaN) == Point(1,2,NaN)`

**Comment by brendan-ward** [comment]
> does that mean you would be OK with the behaviour change to consider NaNs in the same location in the coordinate sequence as equal?

Good to call me on it; I intentionally sidestepped answering that question.  😄   
(was only suggesting comments to better indicate this direction)

I personally haven't relied on this behavior and thus would not be impacted either way (so take my perspective with a grain of salt), and find NaNs to be challenging in coordinate sequences throughout the codebase (I always go back to the comment in [pygeos #306](https://github.com/pygeos/pygeos/issues/306#issuecomment-790143192) of `NaNs make everything bad, don't let them in`).  I think what you have done here is _reasonable_ in that they are functionally equivalent if they occur at exactly the same position even though NaNs are not themselves comparable.

So long as this is clearly noted in the changelog, that hopefully avoids surprises for users, and I really hope very few if any are relying on specific comparisons where NaNs suddenly switch from incomparable (equals is false) to equivalent if functionally equal.

**Comment by jorisvandenbossche** [comment]
Yes, fully agreed on that ideally you never have NaNs in coordinate sequences to start with, and we can improve APIs to prevent users from getting that. 
However, _if_ (for whatever reason) you have NaNs, one should be able to test if two geometries are identical (including their NaNs) in some way. The question then of course is whether `__eq__` needs to be this way. If we add an `equals_identical()` function, people can also use that.

I also don't have real-world experience to base an opinion on, but I have the feeling it makes sense to let `__eq__` be this "identical" check, since you have `equals()` for spatial equality and `equals_exact()` if you want to be able to specify some tolerance. 
Whatever we choose, it's always somewhat ambiguous what `__eq__` exactly does, because of it being an operator and not a function (also right now, users don't necessarily know whether it calls equals() or equals_exact())

**Comment by idanmiara** [comment]
> _if_ (for whatever reason) you have NaNs, one should be able to test if two geometries are identical (including their NaNs) in some way. The question then of course is whether `__eq__` needs to be this way.

(Hope I understood correctly) I strongly agree that `__eq__` needs to return `True` on two geometries that are equal and have NaNs in the same locations, like `Point(1,2,NaN) == Point(1,2,NaN)`

File: src/pygeos.c [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    free(buf1);
+    free(buf2);
+    return 2;
+  }
+
+  ret = 1;
+  for (i = 0; i < n1 * dims1; i++) {
+    const double a = buf1[i];
+    const double b = buf2[i];
+    if (a != b && !(npy_isnan(a) && npy_isnan(b))) {
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply src/pygeos.c:68]: I think the advantage of the current `if (a != b && !(npy_isnan(a) && npy_isnan(b)))` is that for the identical case we already get false for `a != b`, and so don't have to check for NaNs. 
While if we do `if (a == b || (npy_isnan(a) && npy_isnan(b)))`, we would have to check for NaNs for _every_ coordinate pair, even if they are equal.

So I assume that the current implementation should be faster (I could time that to be sure)

**Comment by jorisvandenbossche** [comment]
OK, updated this PR for the review comments. This is now "generally" passing, i.e. for most GEOS versions except for some corner cases where GEOS 3.8 seems to behave differently (for empty geometries, have to further diagnose this). The tests for GEOS main are failing, but those are failing on main as well (and related to https://github.com/libgeos/geos/issues/832).

I also opened a dedicated issue for the API question regarding how NaNs are treated for visibility: https://github.com/shapely/shapely/issues/1775

File: src/pygeos.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (brendan-ward) [reply src/pygeos.c:71]: No need to time it or alter the function, was just a suggestion.

Timing wise, wouldn't `a==b` short circuit out of the OR if true?  That was my assumption, which is why I thought these would be equivalent in timing and only different in readability...

[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply src/pygeos.c:71]: Ah, yes, of course, that should just as well short circuit :facepalm:

**Comment by brendan-ward** [comment]
Thanks for the updates here @jorisvandenbossche .  Can you please add a CHANGELOG entry?

The only part I'm still struggling with is the equality tests with Z nan values that are dependent on the GEOS version, where it returns `False` for GEOS <= 3.9 and GEOS >= 3.12, but `True` for GEOS 3.10-3.11.  The tests document this behavior, which if I'm following correctly, is existing behavior independent of this PR.  But since this doesn't have an obvious docstring where we could note this behavior, I wonder if maybe adding an **Equality** section to `docs/geometry.rst` might be the place to add such a note?  (and also a place to clarify the 3 different types of equality tests: `__eq__`, `equals` (predicate), `equals_exact`).

File: src/pygeos.c [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+/*
+ * Determine pointwise equivalence of two geometries by checking
+ * that the structure, ordering, and values of all vertices are
+ * identical in all dimensions. NaN values are considered to be
+ * equal to other NaN values in the same position within coordinate sequences.
+ * \returns 1 on true, 0 on false, 2 on exception
+ */
+char PyGEOSEqualsIdentical(GEOSContextHandle_t ctx, const GEOSGeometry* geom1,
+                           const GEOSGeometry* geom2) {
+  return equals_identical(ctx, geom1, geom2);
```
Comment (idanmiara) [inline src/pygeos.c:251]: What is the purpose of having both `PyGEOSEqualsIdentical` and `equals_identical` with the same signature when `PyGEOSEqualsIdentical` just calls `equals_identical` ?
Maybe use alias instead if you wanted two names for the same function? https://stackoverflow.com/questions/40968732/how-to-give-alias-for-a-function-in-c

File: shapely/tests/test_geometry.py [hunk 3]
```diff
@@ -244,19 +244,11 @@ def test_set_unique(geom):
...
-    assert line_string_nan != line_string_nan
-
-
-def test_neq_nan():
-    assert not (line_string_nan == line_string_nan)
-
-
 def test_set_nan():
     # As NaN != NaN, you can have multiple "NaN" points in a set
     # set([float("nan"), float("nan")]) also returns a set with 2 elements
```
Comment (idanmiara) [inline shapely/tests/test_geometry.py:249]:
```suggestion
    # This is because "NaN" coordinates in a geometry are considered as equal.
```

File: shapely/tests/test_geometry.py [hunk 4]
```diff
@@ -244,19 +244,11 @@ def test_set_unique(geom):
...
     a = set(shapely.linestrings([[[np.nan, np.nan], [np.nan, np.nan]]] * 10))
-    assert len(a) == 10  # different objects: NaN != NaN
+    assert len(a) == 1  # different objects: NaN != NaN
```
Comment (idanmiara) [inline shapely/tests/test_geometry.py:251]:
```suggestion
    assert len(a) == 1  # same objects: NaN == NaN
 ```

**Comment by jorisvandenbossche** [comment]
This is ready for another review.

File: src/pygeos.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply src/pygeos.c:70]: FWIW, I didn't update for this comment because "not equal until proven equal" actually doesn't work with `break`ing early, as you need to check until the last coordinate to be sure they all are equal.

File: src/pygeos.c [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  coord_seq1 = GEOSGeom_getCoordSeq_r(ctx, geom1);
+  if (coord_seq1 == NULL) {
+    return 2;
+  }
+  coord_seq2 = GEOSGeom_getCoordSeq_r(ctx, geom2);
+  if (coord_seq2 == NULL) {
+    return 2;
+  }
+
+  return coordseq_identical(ctx, coord_seq1, coord_seq2, (int)hasZ1);
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply src/pygeos.c:110]: Merged the two functions into one, but kept the currrent name `equals_identical_simple` (that's consistent with the naming we used in the get_coordinates code)

**Comment by jorisvandenbossche** [comment]
@mwtoews I added some minimal tests for (in)equality with M coordinates

File: src/pygeos.c [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+  if ((type1 == 0) || (type1 == 1) || (type1 == 2)) {
+    return equals_identical_simple(ctx, geom1, geom2);
+  } else if (type1 == 3) {
+    return equals_identical_polygon(ctx, geom1, geom2);
+  } else if ((type1 >= 4) && (type1 <= 7)) {
+    return equals_identical_collection(ctx, geom1, geom2);
+  } else {
+    return 2;
+  }
```
Comment (mwtoews) [inline src/pygeos.c:230]: Is it more readable using enums? And what about a switch?
```suggestion
  switch (type1) {
    case GEOS_POINT:
    case GEOS_LINESTRING:
    case GEOS_LINEARRING:
      return equals_identical_simple(ctx, geom1, geom2);
    case GEOS_POLYGON:
      return equals_identical_polygon(ctx, geom1, geom2);
    case GEOS_MULTIPOINT:
    case GEOS_MULTILINESTRING:
    case GEOS_MULTIPOLYGON:
    case GEOS_GEOMETRYCOLLECTION:
      return equals_identical_collection(ctx, geom1, geom2);
    default:
      return 2;
  }
```

**Review by mwtoews — COMMENTED** [review]
Looks good! I have a few suggestions which you can take or leave.

File: src/pygeos.c (hunk 6 above)
Comment (jorisvandenbossche) [reply src/pygeos.c:230]: Yes, that's indeed nicer

**Comment by jorisvandenbossche** [comment]
@mwtoews thanks for the review!

File: src/geos.h [hunk 7]
```diff
@@ -203,6 +203,6 @@ extern enum ShapelyErrorCode coordseq_from_buffer(GEOSContextHandle_t ctx,
                                                   npy_intp cs2,
                                 GEOSCoordSequence** coord_seq);
 extern int coordseq_to_buffer(GEOSContextHandle_t ctx, const GEOSCoordSequence* coord_seq,
-                              double* buf, unsigned int size, unsigned int dims);
+                              double* buf, unsigned int size, int hasZ, int hasM);
```
[the remark this answers is not in the document]
Comment (mwtoews) [reply src/geos.h:206]: This can be left for now and further discussed in #1648 as it can change later.

But to clarify my suggestion a bit more, the argument naming convention "foo_bar" is used in `src/*.h` header files, and even skimming header files from [`numpy.get_include()`](https://github.com/numpy/numpy/tree/main/numpy/_core/include/numpy) too. Data types and internal variables are fine to take whatever convention they wish (i.e. "fooBar").
